File: 27-Interpreter/src/cpp/main.cpp
#include <iostream>
#include <memory>
#include "Expression.h"
#include "Context.h"
#include "Variable.h"
#include "AndExpression.h"
#include "OrExpression.h"
#include "NotExpression.h"

static std::unique_ptr<Expression> build_rule() {
    // (A OR B) AND (NOT C)
    return std::make_unique<AndExpression>(
        std::make_unique<OrExpression>(
            std::make_unique<Variable>("A"),
            std::make_unique<Variable>("B")),
        std::make_unique<NotExpression>(
            std::make_unique<Variable>("C")));
}

static void evaluate(const Expression& rule, bool a, bool b, bool c) {
    Context context;
    context.set("A", a);
    context.set("B", b);
    context.set("C", c);
    std::cout << "  A=" << a << ", B=" << b << ", C=" << c
              << " -> " << rule.interpret(context) << "\n";
}

int main() {
    std::cout << "=== Interpreter Pattern Demo (C++) ===\n\n";
    std::cout << "Rule: (A OR B) AND (NOT C)\n\n";

    auto rule = build_rule();

    evaluate(*rule, true, false, false);
    evaluate(*rule, false, false, false);
    evaluate(*rule, true, true, true);

    return 0;
}

File: 27-Interpreter/src/rust/main.rs
mod expression;
mod context;
mod variable;
mod constant;
mod and_expression;
mod or_expression;
mod not_expression;

use expression::Expression;
use context::Context;
use variable::Variable;
use and_expression::AndExpression;
use or_expression::OrExpression;
use not_expression::NotExpression;

fn evaluate(rule: &dyn Expression, a: bool, b: bool, c: bool) {
    let mut context = Context::new();
    context.set("A", a);
    context.set("B", b);
    context.set("C", c);
    println!("  A={}, B={}, C={} -> {}", a, b, c, rule.interpret(&context));
}

fn main() {
    println!("=== Interpreter Pattern Demo (Rust) ===\n");

    // Build the rule tree for: (A OR B) AND (NOT C)
    let rule = AndExpression::new(
        Box::new(OrExpression::new(
            Box::new(Variable::new("A")),
            Box::new(Variable::new("B")),
        )),
        Box::new(NotExpression::new(Box::new(Variable::new("C")))),
    );

    println!("Rule: (A OR B) AND (NOT C)\n");

    evaluate(&rule, true, false, false);
    evaluate(&rule, false, false, false);
    evaluate(&rule, true, true, true);
}
